packages/material-ui/src/InputBase/Textarea.js

```diff
@@ -1,246 +1,168 @@
 import React from 'react';
 import PropTypes from 'prop-types';
-import clsx from 'clsx';
+import useForkRef from '../utils/useForkRef';
 import debounce from 'debounce'; // < 1kb payload overhead when lodash/debounce is > 3kb.
-import EventListener from 'react-event-listener';
-import withStyles from '../styles/withStyles';
-import { setRef } from '../utils/reactHelpers';
 
-const ROWS_HEIGHT = 19;
-
-export const styles = {
-  /* Styles applied to the root element. */
-  root: {
-    position: 'relative', // because the shadow has position: 'absolute',
-    width: '100%',
-  },
-  textarea: {
-    width: '100%',
-    height: '100%',
-    resize: 'none',
-    font: 'inherit',
-    padding: 0,
-    cursor: 'inherit',
-    boxSizing: 'border-box',
-    lineHeight: 'inherit',
-    border: 'none',
-    outline: 'none',
-    background: 'transparent',
-  },
+const styles = {
+  /* Styles applied to the shallow textarea element. */
   shadow: {
-    // Overflow also needed to here to remove the extra row
-    // added to textareas in Firefox.
-    overflow: 'hidden',
     // Visibility needed to hide the extra text area on iPads
     visibility: 'hidden',
+    // Remove from the content flow
     position: 'absolute',
-    height: 'auto',
-    whiteSpace: 'pre-wrap',
+    // Ignore the scrollbar width
+    overflow: 'hidden',
+    height: '0',
   },
 };
 
+function getStyleValue(computedStyle, property) {
+  return parseInt(computedStyle[property], 10) || 0;
+}
+
+const useEnhancedEffect = typeof window !== 'undefined' ? React.useLayoutEffect : React.useEffect;
+
 /**
  * @ignore - internal component.
+ *
+ * To make public in v4+.
  */
-class Textarea extends React.Component {
-  constructor(props) {
-    super();
-    this.isControlled = props.value != null;
-    // <Input> expects the components it renders to respond to 'value'
-    // so that it can check whether they are filled.
-    this.value = props.value || props.defaultValue || '';
-    this.state = {
-      height: Number(props.rows) * ROWS_HEIGHT,
-    };
-
-    if (typeof window !== 'undefined') {
-      this.handleResize = debounce(() => {
-        this.syncHeightWithShadow();
-      }, 166); // Corresponds to 10 frames at 60 Hz.
-    }
-  }
-
-  componentDidMount() {
-    this.syncHeightWithShadow();
-  }
+const Textarea = React.forwardRef(function Textarea(props, ref) {
+  const { onChange, rowsMax, rowsMin, style, value, ...other } = props;
 
-  componentDidUpdate() {
-    this.syncHeightWithShadow();
-  }
+  const { current: isControlled } = React.useRef(value != null);
+  const inputRef = React.useRef();
+  const [state, setState] = React.useState({});
+  const shallowRef = React.useRef();
+  const handleRef = useForkRef(ref, inputRef);
 
-  componentWillUnmount() {
-    this.handleResize.clear();
-  }
+  const syncHeight = () => {
```

Comment:
This should be memoized and only attached to window.onResize when it changes. Right now we re-attach on every render.